File: 3234-count-the-number-of-substrings-with-dominant-ones.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        '''
        If we have 1 to 3 ones, then we at most can have 1 zero
        4 ones, at most is 2 zeros

        In reverse, if 1 zero, we will need 1 to 3 ones.

        Max number of zeros for s of length `n` = sqrt(n)
        We can build a list of zeros, then for each index i in s, we try to check the number of zeros

        Time complexity: O(n * sqrt(n))
        '''

        right = 0
        n = len(s)
        next_zero = [-1] * n

        prev = n
        for i in range(n-1, -1, -1):
            next_zero[i] = prev
            if s[i] == '0':
                prev = i

        res = 0

        for left in range(n):
            zeroes = 1 if s[left] == '0' else 0
            right = left

            while zeroes ** 2 <= n:
                next = next_zero[right]
                ones = next - left - zeroes
                if ones >= zeroes ** 2:
                    res += min(next - right, ones - (zeroes**2) + 1)
                zeroes += 1
                right = next
                if right == n:
                    break

        return res


sol = Solution()
print(f'output: {sol.numberOfSubstrings("00011")}, expected: 5')
print(f'output: {sol.numberOfSubstrings("101101")}, expected: 16')

File: 3234-count-the-number-of-substrings-with-dominant-ones.java
class CountTheNumberOfSubstringsWithDominantOnes {

    public int numberOfSubstrings(String s) {
        int right = 0;
        int n = s.length();
        int[] nextZero = new int[n];

        int prev = n;
        for (int i = n - 1; i > -1; i--) {
            nextZero[i] = prev;
            if (s.charAt(i) == '0') {
                prev = i;
            }
        }

        int res = 0;
        for (int left = 0; left < n; left++) {
            int zeroes = 0, ones = 0;
            if (s.charAt(left) == '0')
                zeroes = 1;
            right = left;

            while (zeroes * zeroes <= n) {
                int next = nextZero[right];
                ones = next - left - zeroes;
                if (ones >= zeroes * zeroes) {
                    res += Math.min(next - right, ones - (zeroes * zeroes) + 1);
                }
                zeroes++;
                right = next;
                if (right == n) {
                    break;
                }
            }
        }
        return res;
    }

    public static void main(String[] args) {
        CountTheNumberOfSubstringsWithDominantOnes sol = new CountTheNumberOfSubstringsWithDominantOnes();
        System.out.println("Output: " + sol.numberOfSubstrings("00011") + ", expected: 5");
        System.out.println("Output: " + sol.numberOfSubstrings("101101") + ", expected: 16");
    }
}
